File: packages/javascript/test/undefined-key.test.ts
import assert from 'node:assert';
import { join } from 'node:path';
import { describe, it } from 'node:test';
import {
	analyzeTranslations,
	CHECK_CODE,
	checkTranslationFiles,
	findUnusedKeys,
	resolveOptions
} from '../dist/index.js';

const localesPath = join('test', 'samples', 'undefined-key', 'locales');

const sourcePath = join('test', 'samples', 'undefined-key', 'src');

const KEYS = ['desc.hello', 'attr.folder', 'item_one', 'item_other'];

const scan = (options = {}) =>
	findUnusedKeys(sourcePath, KEYS, resolveOptions({ target: 'en', ...options }).options);

describe('findUnusedKeys, on the keys the source asks for', () => {
	it('reports a key no language file defines', async () => {
		const { undefinedKeys } = await scan();

		assert.deepStrictEqual(
			undefinedKeys.map((usage) => usage.key),
			['attr.missing']
		);
		assert.ok(undefinedKeys[0].file?.endsWith('app.ts'));
	});

	it('reads a key through the namespace written in front of it', async () => {
		const { undefinedKeys } = await scan();

		// `t('common:desc.hello')` names a namespace and a key that does exist.
		assert.strictEqual(
			undefinedKeys.some((usage) => usage.key.includes('common')),
			false
		);
	});

	it('leaves a key built at run time alone', async () => {
		const { undefinedKeys } = await scan();

		assert.strictEqual(
			undefinedKeys.some((usage) => usage.key.startsWith('error.')),
			false
		);
	});

	it('counts a key reached through a prefix as defined', async () => {
		const { undefinedKeys } = await scan();

		// `t('folder')` resolves to `attr.folder` through a bound prefix.
		assert.strictEqual(
			undefinedKeys.some((usage) => usage.key === 'folder'),
			false
		);
	});

	it('counts a plural key asked for by its base as defined', async () => {
		const { undefinedKeys, unusedKeys } = await scan();

		// The source writes `t('item')`; the runtime picks `item_one`.
		assert.strictEqual(
			undefinedKeys.some((usage) => usage.key === 'item'),
			false
		);
		assert.deepStrictEqual(unusedKeys, []);
	});

	it('reads the call names the project actually uses', async () => {
		const { undefinedKeys } = await scan({ translateFunctions: ['nothing'] });

		assert.deepStrictEqual(undefinedKeys, []);
	});

	it('does the work only for the check that needs it', async () => {
		const { undefinedKeys } = await scan({ ignoreChecks: [CHECK_CODE.UNDEFINED_KEY] });

		assert.deepStrictEqual(undefinedKeys, []);
	});
});

describe('UNDEFINED_KEY', () => {
	it('reports what the source asks for and the files do not have', async () => {
		const result = await checkTranslationFiles(localesPath, {
			target: 'en',
			source: sourcePath,
			checks: [CHECK_CODE.UNDEFINED_KEY]
		});

		assert.strictEqual(result.issuesByCode.UNDEFINED_KEY?.length, 1);
		assert.strictEqual(result.issuesByCode.UNDEFINED_KEY?.[0].key, 'attr.missing');
		// Not a locale's fault: no language file defines it.
		assert.strictEqual(result.issuesByCode.UNDEFINED_KEY?.[0].locale, '');
	});

	it('reports nothing without a source directory to read', async () => {
		const result = await checkTranslationFiles(localesPath, {
			target: 'en',
			checks: [CHECK_CODE.UNDEFINED_KEY]
		});

		assert.strictEqual(result.issues.length, 0);
	});

	it('accepts an answer worked out elsewhere', () => {
		const result = analyzeTranslations(
			{ locales: { en: { a: 'A' } }, undefinedKeys: [{ key: 'b.c', file: 'app.ts' }] },
			{ target: 'en', checks: [CHECK_CODE.UNDEFINED_KEY] }
		);

		assert.strictEqual(result.issuesByCode.UNDEFINED_KEY?.[0].key, 'b.c');
		assert.strictEqual(result.issuesByCode.UNDEFINED_KEY?.[0].file, 'app.ts');
	});

	it('never fails a run on its own', async () => {
		const result = await checkTranslationFiles(localesPath, {
			target: 'en',
			source: sourcePath,
			checks: [CHECK_CODE.UNDEFINED_KEY]
		});

		assert.strictEqual(result.success, true);
		assert.strictEqual(result.summary.warn, 1);
	});
});

File: packages/dart/test/undefined_key_test.dart
import 'package:chki18n/chki18n.dart';
import 'package:test/test.dart';

const String localesPath = 'test/samples/undefined-key/locales';

const String sourcePath = 'test/samples/undefined-key/src';

const List<String> sampleKeys = ['desc.hello', 'attr.folder', 'item_one', 'item_other'];

Future<Chki18nUsageScan> scan({Chki18nOptions options = const Chki18nOptions()}) => findUnusedKeys(
  sourcePath,
  sampleKeys,
  resolveOptions(options.copyWith(target: options.target ?? 'en')).options,
);

void main() {
  group('findUnusedKeys, on the keys the source asks for', () {
    test('reports a key no language file defines', () async {
      final found = await scan();

      expect([for (final usage in found.undefinedKeys) usage.key], ['attr.missing']);
      expect(found.undefinedKeys.first.file, endsWith('app.ts'));
    });

    test('reads a key through the namespace written in front of it', () async {
      final found = await scan();

      // `t('common:desc.hello')` names a namespace and a key that does exist.
      expect(found.undefinedKeys.any((usage) => usage.key.contains('common')), isFalse);
    });

    test('leaves a key built at run time alone', () async {
      final found = await scan();

      expect(found.undefinedKeys.any((usage) => usage.key.startsWith('error.')), isFalse);
    });

    test('counts a key reached through a prefix as defined', () async {
      final found = await scan();

      // `t('folder')` resolves to `attr.folder` through a bound prefix.
      expect(found.undefinedKeys.any((usage) => usage.key == 'folder'), isFalse);
    });

    test('counts a plural key asked for by its base as defined', () async {
      final found = await scan();

      // The source writes `t('item')`; the runtime picks `item_one`.
      expect(found.undefinedKeys.any((usage) => usage.key == 'item'), isFalse);
      expect(found.unusedKeys, isEmpty);
    });

    test('reads the call names the project actually uses', () async {
      final found = await scan(options: const Chki18nOptions(translateFunctions: ['nothing']));

      expect(found.undefinedKeys, isEmpty);
    });

    test('does the work only for the check that needs it', () async {
      final found = await scan(
        options: const Chki18nOptions(ignoreChecks: [Chki18nCheckCode.undefinedKey]),
      );

      expect(found.undefinedKeys, isEmpty);
    });
  });

  group('UNDEFINED_KEY', () {
    test('reports what the source asks for and the files do not have', () async {
      final result = await checkTranslationFiles(
        path: localesPath,
        options: const Chki18nOptions(
          target: 'en',
          source: sourcePath,
          checks: [Chki18nCheckCode.undefinedKey],
        ),
      );

      expect(result.of(Chki18nCheckCode.undefinedKey).length, 1);
      expect(result.of(Chki18nCheckCode.undefinedKey).first.key, 'attr.missing');
      // Not a locale's fault: no language file defines it.
      expect(result.of(Chki18nCheckCode.undefinedKey).first.locale, '');
    });

    test('reports nothing without a source directory to read', () async {
      final result = await checkTranslationFiles(
        path: localesPath,
        options: const Chki18nOptions(target: 'en', checks: [Chki18nCheckCode.undefinedKey]),
      );

      expect(result.issues, isEmpty);
    });

    test('accepts an answer worked out elsewhere', () {
      final result = analyzeTranslations(
        const Chki18nInput(
          locales: {
            'en': {'a': 'A'},
          },
          undefinedKeys: [Chki18nKeyUsage(key: 'b.c', file: 'app.ts')],
        ),
        options: const Chki18nOptions(target: 'en', checks: [Chki18nCheckCode.undefinedKey]),
      );

      expect(result.of(Chki18nCheckCode.undefinedKey).first.key, 'b.c');
      expect(result.of(Chki18nCheckCode.undefinedKey).first.file, 'app.ts');
    });

    test('never fails a run on its own', () async {
      final result = await checkTranslationFiles(
        path: localesPath,
        options: const Chki18nOptions(
          target: 'en',
          source: sourcePath,
          checks: [Chki18nCheckCode.undefinedKey],
        ),
      );

      expect(result.success, isTrue);
      expect(result.summary.warn, 1);
    });
  });
}
